nuxeo-core/nuxeo-core-storage/src/main/java/org/nuxeo/ecm/core/storage/ExpressionEvaluator.java

```diff
@@ -270,26 +272,34 @@ protected Boolean walkIsTrashed(Operator op, Operand rvalue) {
         if (op != Operator.EQ && op != Operator.NOTEQ) {
             throw new QueryParseException(NXQL.ECM_ISTRASHED + " requires = or <> operator");
         }
-        long v;
-        if (!(rvalue instanceof IntegerLiteral)
-                || ((v = ((IntegerLiteral) rvalue).value) != 0 && v != 1)) {
-            throw new QueryParseException(NXQL.ECM_ISTRASHED + " requires literal 0 or 1 as right argument");
-        }
-        Reference ref;
-        Literal val;
         TrashService trashService = Framework.getService(TrashService.class);
-        if (trashService.hasFeature(TRASHED_STATE_IS_DEDICATED_PROPERTY)) {
-            ref = new Reference(NXQL.ECM_ISTRASHED);
-            val = new BooleanLiteral(true); // give true to match equalsDeleted mechanism
+        if (trashService.hasFeature(TRASHED_STATE_IS_DEDUCED_FROM_LIFECYCLE)) {
+            return walkIsTrashed(new Reference(NXQL.ECM_LIFECYCLESTATE), op, rvalue,
+                    new StringLiteral(LifeCycleConstants.DELETED_STATE));
+        } else if (trashService.hasFeature(TRASHED_STATE_IN_MIGRATION)) {
+            Boolean lifeCycleTrashed = walkIsTrashed(new Reference(NXQL.ECM_LIFECYCLESTATE), op, rvalue,
+                    new StringLiteral(LifeCycleConstants.DELETED_STATE));
+            Boolean propertyTrashed = walkIsTrashed(new Reference(NXQL.ECM_ISTRASHED), op, rvalue,
+                    new BooleanLiteral(true));
+            return or(lifeCycleTrashed, propertyTrashed);
+        } else if (trashService.hasFeature(TRASHED_STATE_IS_DEDICATED_PROPERTY)) {
+            return walkIsTrashed(new Reference(NXQL.ECM_ISTRASHED), op, rvalue, new BooleanLiteral(true));
         } else {
-            ref = new Reference(NXQL.ECM_LIFECYCLESTATE);
-            val = new StringLiteral(LifeCycleConstants.DELETED_STATE);
+            throw new UnsupportedOperationException("TrashService is in an unknown state");
+        }
+    }
+
+    protected Boolean walkIsTrashed(Reference ref, Operator op, Operand initialRvalue, Literal deletedRvalue) {
+        long v;
+        if (!(initialRvalue instanceof IntegerLiteral)
+                || ((v = ((IntegerLiteral) initialRvalue).value) != 0 && v != 1)) {
+            throw new QueryParseException(NXQL.ECM_ISTRASHED + " requires literal 0 or 1 as right argument");
```

Comment:
not easy to read and duplicate in MongoDBQueryBuilder could be hided in method

Reply:
`MessageFormat.format("{0} requires literal 0 or 1 as right argument", ECM_ISTRASHED);`

Reply:
`MessageFormat` is overkill for a simple 2-string concatenation

Reply:
I didn't change that, as it was already there.